### DISCUSSION TITLE
nx release appears not to update peerDependencies of monorepo packages

### DISCUSSION THREAD

**Discussion by zpydee** [nrwl/nx#23388] [Feature Requests]
### Current Behavior

when running nx release, it updates the version numbers of my dependencies, but not peerDependencies

### Expected Behavior

should update peerDependencies

### Steps to Reproduce

1. create publishable package a (mine is a js lib)
2. create publishable package b (mine is a vite react lib)
3. add package a as a peerDependency in (b)
4. run nx release

### Nx Report

```shell
Node   : 20.12.1
OS     : darwin-arm64
pnpm   : 8.15.6

nx (global)        : 18.2.4
nx                 : 18.2.4
@nx/js             : 18.2.4
@nx/jest           : 18.2.4
@nx/eslint         : 18.2.4
@nx/workspace      : 18.2.4
@nx/devkit         : 18.2.4
@nx/eslint-plugin  : 18.2.4
@nx/next           : 18.2.4
@nx/plugin         : 18.2.4
@nx/react          : 18.2.4
@nx/storybook      : 18.2.4
@nrwl/tao          : 18.2.4
@nx/vite           : 18.2.4
@nx/web            : 18.2.4
typescript         : 5.4.5
---------------------------------------
Registered Plugins:
@nx/next/plugin
@nx/storybook/plugin
@nx/vite/plugin
@nx/eslint/plugin
pnpm -
```
```


### Failure Logs

_No response_

### Package Manager Version

pnpm@8.15.6

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

_No response_

**Comment by de-don** [comment]
Faced the same issue. I have two libraries:
- `parent-lib@1.0.0`
- `child-lib@1.0.0` - peerDependency `"parent-lib: "^1.0.0"`

Child lib has peerDependency `"parent-lib: "^1.0.0"`

When I commit some breaking change to parent lib (and update child lib as well because it's affected), I get next results:
- `parent-lib@2.0.0`
- `child-lib@2.0.0` - peerDependency `"parent-lib: "^1.0.0"`

It's a bit strange behaviour, especially when you run `version` and `publish` on CI (spoiler it will publish child-lib with wrong depencency).

**Comment by ryanbas21** [comment]
+1 to this issue. I see this also
My structure is similar
```
packages/
   - a
   - b 
      - peer dep on A
   - c
      - peer dep on A
```

I have to manually update the package.json of the peer dep

**Comment by gperdomor** [comment]
Same here, I notice that the versions are changed in `dist` folder but not in the sources files inside `packages` directory, at least in version 18.3.3

Sample output:
```
 NX   Running release version for project: core

core 🔍 Reading data for package "@nx-tools/core" from dist/packages/core/package.json
core 📄 Using the current version 6.0.0-alpha.2 already resolved from disk fallback.
core ✍️  New version 6.0.0-alpha.3 written to dist/packages/core/package.json
core ✍️  Applying new version 6.0.0-alpha.3 to 5 packages which depend on core

UPDATE dist/packages/container-metadata/package.json [dry-run]

    "name": "@nx-tools/container-metadata",
-   "version": "6.0.0-alpha.2",
+   "version": "6.0.0-alpha.3",
    "type": "commonjs",

    "dependencies": {
-     "@nx-tools/ci-context": "6.0.0-alpha.2",
-     "@nx-tools/core": "6.0.0-alpha.2",
+     "@nx-tools/ci-context": "6.0.0-alpha.3",
+     "@nx-tools/core": "6.0.0-alpha.3",
      "@renovate/pep440": "1.0.0",
 ```

**Comment by adamjosefus** [comment]
Same. If I changed `peerDependencies` to `dependencies` then version was resolved. Eg. from `workspace:*` to `0.1.1`

**Comment by fahslaj** [comment]
To avoid being brittle, peer dependencies are often ranges that require intentionality and manual intervention to update appropriately, and they are often a breaking change to the package. This is why Nx Release doesn't update them automatically at the moment. If you have specific exact versions, I would suggest using `dependencies` over `peerDependencies`.

However, with enough support and carefully thought out use cases, this is something we could consider as an opt-in behavior for Nx Release. Please continue to share your use cases and justification for automatically-updated peer dependencies. This type of discussion and feedback is what will make a future implementation of the feature as well thought out and valuable as possible!

**Reply by czycha** [reply]
In our particular case, we have a package with a peer that is tightly coupled—they should always be installed at the same version as each other (kinda like Nx and its plugins). I think in our case, a satisfying answer  would be to check whether the peer dependency version is exactly equal with no approximations/ranges (ex. `"@my/peer-dep": "1.0.0"`). That might even be applicable outside of peer deps as well in case a direct dependency only cares to match any minor/patch version of another project using `^x.y.z`.

**Reply by Badisi** [reply]
In my case, I have a monorepo with multiple packages and interdependencies.

All my packages are symlinked thanks to `package.json#workspaces`:

```md
- projects/
    - a/
        - package.json
            "name": "a",
            "version": "1.2.0"
    - b/
        - package.json
            "name": "b",
            "version": "1.0.0",
            "peerDependencies": {
                "a": "1.2.0"
            }

- package.json
    "workspaces": [
        "projects/a",
        "projects/b"
    ]
```

Now, let's say a new version of package `a` is released (e.g.: `1.3.0`).

Currently, the version in `projects/a/package.json` will be updated to `1.3.0`.
But not the version in `projects/b/package.json#peerDependencies#a`.

This results in the next `npm install` failing because `a@1.2.0` no longer exists locally.

-----

So my personal concerns for keeping peer dependencies in sync are:
1. Ensuring `npm install` is not broken
2. Knowing when one of my package has a peer dependency update available and needs to be released

For point 2, I have created a [script](https://github.com/DSI-HUG/ngx-components/blob/main/scripts/release.ts#L128) that automatically syncs every peer dependencies in the repo, and makes special commits that appear in the release changelog so that I know when a package needs to be released (ie. in the previous example, the release dry-run of `b` would indicates that `a` was updated and that `b` also needs to be released).
